// ionic-framework/src/es5/common/ionic/components/tap-click/activator.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _utilDom = require('../../util/dom');

var Activator = (function () {
    function Activator(app, config) {
        _classCallCheck(this, Activator);

        this.app = app;
        this.queue = [];
        this.active = [];
        this.clearStateTimeout = 180;
        this.clearAttempt = 0;
        this.activatedClass = config.setting('activatedClass') || 'activated';
        this.x = 0;
        this.y = 0;
    }

    _createClass(Activator, [{
        key: 'downAction',
        value: function downAction(ev, activatableEle, pointerX, pointerY, callback) {
            var _this = this;

            // the user just pressed down
            if (this.disableActivated(ev)) return;
            // remember where they pressed
            this.x = pointerX;
            this.y = pointerY;
            // queue to have this element activated
            this.queue.push(activatableEle);
            (0, _utilDom.raf)(function () {
                var activatableEle = undefined;
                for (var i = 0; i < _this.queue.length; i++) {
                    activatableEle = _this.queue[i];
                    if (activatableEle && activatableEle.parentNode) {
                        _this.active.push(activatableEle);
                        activatableEle.classList.add(_this.activatedClass);
                    }
                }
                _this.queue = [];
            });
        }
    }, {
        key: 'upAction',
        value: function upAction() {
            var _this2 = this;

            // the user was pressing down, then just let up
            setTimeout(function () {
                _this2.clearState();
            }, this.clearStateTimeout);
        }
    }, {
        key: 'clearState',
        value: function clearState() {
            // all states should return to normal
            if ((!this.app.isEnabled() || this.app.isTransitioning()) && this.clearAttempt < 30) {
                // the app is actively disabled, so don't bother deactivating anything.
                // this makes it easier on the GPU so it doesn't have to redraw any
                // buttons during a transition. This will retry in XX milliseconds.
                ++this.clearAttempt;
                this.upAction();
            } else {
                // not actively transitioning, good to deactivate any elements
                this.deactivate();
                this.clearAttempt = 0;
            }
        }
    }, {
        key: 'deactivate',
        value: function deactivate() {
            // remove the active class from all active elements
            for (var i = 0; i < this.active.length; i++) {
                this.active[i].classList.remove(this.activatedClass);
            }
            this.queue = [];
            this.active = [];
        }
    }, {
        key: 'disableActivated',
        value: function disableActivated(ev) {
            var targetEle = ev.target;
            for (var x = 0; x < 4; x++) {
                if (!targetEle) break;
                if (targetEle.hasAttribute('disable-activated')) return true;
                targetEle = targetEle.parentElement;
            }
            return false;
        }
    }]);

    return Activator;
})();

exports.Activator = Activator;
