// ionic-framework/src/es6/ionic/components/nav/nav-router.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { Directive, ElementRef, DynamicComponentLoader, Attribute } from 'angular2/angular2';
import { RouterOutlet, Router, Instruction } from 'angular2/router';
import { Nav } from './nav';
/**
 * TODO
 */
export let NavRouter = class extends RouterOutlet {
    /**
     * TODO
     * @param {ElementRef} _elementRef  TODO
     * @param {DynamicComponentLoader} _loader  TODO
     * @param {Router} _parentRouter  TODO
     * @param {string} nameAttr  Value of the element's 'name' attribute
     * @param {Nav} nav  TODO
     */
    constructor(_elementRef, _loader, _parentRouter, nameAttr, nav) {
        super(_elementRef, _loader, _parentRouter, nameAttr);
        // Nav is Ionic's NavController, which we injected into this class
        this.nav = nav;
        // register this router with Ionic's NavController
        // Ionic's NavController will call this NavRouter's "stateChange"
        // method when the NavController has...changed its state
        nav.registerRouter(this);
    }
    /**
     * @private
     * TODO
     * @param {ComponentInstruction} instruction  TODO
     */
    activate(nextInstruction) {
        var previousInstruction = this._currentInstruction;
        this._currentInstruction = nextInstruction;
        var componentType = nextInstruction.componentType;
        var childRouter = this._parentRouter.childRouter(componentType);
        // prevent double navigations to the same view
        var lastView = this.nav.last();
        if (lastView && lastView.componentType === componentType && lastView.params.data === nextInstruction.params) {
            return Promise.resolve();
        }
        // tell the NavController which componentType, and it's params, to navigate to
        return this.nav.push(componentType, nextInstruction.params);
    }
    reuse(nextInstruction) {
        return Promise.resolve();
    }
    /**
     * TODO
     * @param {TODO} type  TODO
     * @param {TODO} viewCtrl  TODO
     */
    stateChange(type, viewCtrl) {
        // stateChange is called by Ionic's NavController
        // type could be "push" or "pop"
        // viewCtrl is Ionic's ViewController class, which has the properties "componentType" and "params"
        // only do an update if there's an actual view change
        if (!viewCtrl || this._activeViewId === viewCtrl.id)
            return;
        this._activeViewId = viewCtrl.id;
        // get the best PathRecognizer for this view's componentType
        let pathRecognizer = this.getPathRecognizerByComponent(viewCtrl.componentType);
        if (pathRecognizer) {
            // generate a componentInstruction from the view's PathRecognizer and params
            let componentInstruction = pathRecognizer.generate(viewCtrl.params.data);
            // create an Instruction from the componentInstruction
            let instruction = new Instruction(componentInstruction, null);
            // update the browser's URL
            this._parentRouter.navigateInstruction(instruction);
        }
    }
    /**
     * TODO
     * @param {TODO} componentType  TODO
     * @returns {TODO} TODO
     */
    getPathRecognizerByComponent(componentType) {
        // given a componentType, figure out the best PathRecognizer to use
        let rules = this._parentRouter.registry._rules;
        let pathRecognizer = null;
        rules.forEach((rule) => {
            pathRecognizer = rule.matchers.find((matcherPathRecognizer) => {
                return (matcherPathRecognizer.handler.componentType === componentType);
            });
        });
        return pathRecognizer;
    }
};
NavRouter = __decorate([
    Directive({
        selector: 'ion-nav'
    }),
    __param(3, Attribute('name')), 
    __metadata('design:paramtypes', [(typeof ElementRef !== 'undefined' && ElementRef) || Object, (typeof DynamicComponentLoader !== 'undefined' && DynamicComponentLoader) || Object, (typeof Router !== 'undefined' && Router) || Object, String, (typeof Nav !== 'undefined' && Nav) || Object])
], NavRouter);
