// ionic-framework/src/es5/common/ionic/platform/geolocation/geolocation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj["default"] = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _rx = require('rx');

var Rx = _interopRequireWildcard(_rx);

var _plugin = require('../plugin');

/**
 * Get geolocation data.
 *
 * @usage
 * ```js
 * Geolocation.getCurrentPosition().then((resp) => {
 *  //resp.coords.latitude
 *  //resp.coords.longitude
 * })
 *
 * let watch = Geolocation.watchPosition();
 * watch.source.subscribe((data) => {
 *  //data.coords.latitude
 *  //data.coords.longitude
 * })
 * ```
 */
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2:
            return decorators.reduceRight(function (o, d) {
                return d && d(o) || o;
            }, target);
        case 3:
            return decorators.reduceRight(function (o, d) {
                return (d && d(target, key), void 0);
            }, void 0);
        case 4:
            return decorators.reduceRight(function (o, d) {
                return d && d(target, key, o) || o;
            }, desc);
    }
};
var __metadata = undefined && undefined.__metadata || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var Geolocation = (function () {
    function Geolocation() {
        _classCallCheck(this, Geolocation);
    }

    _createClass(Geolocation, null, [{
        key: "getCurrentPosition",

        /**
         * Get the current GPS location.
         */
        value: function getCurrentPosition(options) {
            return new Promise(function (resolve, reject) {
                navigator.geolocation.getCurrentPosition(function (result) {
                    resolve(result);
                }, function (err) {
                    reject(err);
                }, options);
            });
        }

        /**
         * Watch for location changes.
         */
    }, {
        key: "watchPosition",
        value: function watchPosition(options) {
            var watchID = undefined;
            var source = Rx.Observable.create(function (observer) {
                watchID = navigator.geolocation.watchPosition(function (result) {
                    observer.onNext(result);
                }, function (err) {
                    observer.onError(err, observer);
                }, options);
            });
            return {
                source: source,
                watchID: watchID,
                clear: function clear() {
                    navigator.geolocation.clearWatch(watchID);
                }
            };
        }

        /**
         * Clear a specific watch by watch ID. Generally, you'll call
         * clear() on the returned watch from `getCurrentPosition` or `watchPosition` above.
         */
    }, {
        key: "clearWatch",
        value: function clearWatch(watchID) {
            return navigator.geolocation.clearWatch(watchID);
        }
    }]);

    return Geolocation;
})();
exports.Geolocation = Geolocation;
exports.Geolocation = Geolocation = __decorate([(0, _plugin.NativePlugin)({
    name: 'Geolocation',
    platforms: ['ios', 'android', 'web'],
    engines: {
        cordova: 'cordova-plugin-geolocation'
    }
}), __metadata('design:paramtypes', [])], Geolocation);
