// ionic-framework/src/es6/ionic/components/menu/menu-close.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, ElementRef } from 'angular2/angular2';
import { Ion } from '../ion';
import { IonicApp } from '../app/app';
/**
* TODO
*/
export let MenuClose = class extends Ion {
    constructor(app, elementRef) {
        super(elementRef, null);
        this.app = app;
    }
    close() {
        let menu = this.app.getComponent(this.menuClose || 'menu');
        menu && menu.close();
    }
};
MenuClose = __decorate([
    Directive({
        selector: '[menu-close]',
        properties: [
            'menuClose'
        ],
        host: {
            '(click)': 'close()'
        }
    }), 
    __metadata('design:paramtypes', [(typeof IonicApp !== 'undefined' && IonicApp) || Object, (typeof ElementRef !== 'undefined' && ElementRef) || Object])
], MenuClose);
