// ionic-framework/src/es5/common/ionic/components/radio/radio.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; }

var _angular2Angular2 = require('angular2/angular2');

var _configDecorators = require('../../config/decorators');

var _configConfig = require('../../config/config');

var _ion = require('../ion');

var _listList = require('../list/list');

/**
 * @name ionRadioGroup
 * @description
 * A radio group is a group of radio components.
 *
 * Selecting a radio button in the group unselects all others in the group.
 *
 * New radios can be registered dynamically.
 *
 * See the [Angular 2 Docs](https://angular.io/docs/js/latest/api/forms/) for more info on forms and input.
 *
 * @usage
 * ```html
 * <ion-radio-group ng-control="clientside">
 *
 *   <ion-header>
 *     Clientside
 *   </ion-header>
 *
 *   <ion-radio value="ember">
 *     Ember
 *   </ion-radio>
 *
 *   <ion-radio value="angular1">
 *     Angular 1
 *   </ion-radio>
 *
 *   <ion-radio value="angular2" checked="true">
 *     Angular 2
 *   </ion-radio>
 *
 *   <ion-radio value="react">
 *     React
 *   </ion-radio>
 *
 * </ion-radio-group>
 * ```
*/
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2:
            return decorators.reduceRight(function (o, d) {
                return d && d(o) || o;
            }, target);
        case 3:
            return decorators.reduceRight(function (o, d) {
                return (d && d(target, key), void 0);
            }, void 0);
        case 4:
            return decorators.reduceRight(function (o, d) {
                return d && d(target, key, o) || o;
            }, desc);
    }
};
var __metadata = undefined && undefined.__metadata || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = undefined && undefined.__param || function (paramIndex, decorator) {
    return function (target, key) {
        decorator(target, key, paramIndex);
    };
};
var RadioGroup = (function (_Ion) {
    _inherits(RadioGroup, _Ion);

    /**
     * TODO
     * @param {ElementRef} elementRef  TODO
     * @param {IonicConfig} config  TODO
     * @param {NgControl=} ngControl  TODO
     * @param {QueryList<ListHeader>} headerQuery  TODO
     */

    function RadioGroup(elementRef, config, ngControl, headerQuery) {
        _classCallCheck(this, RadioGroup);

        _get(Object.getPrototypeOf(RadioGroup.prototype), "constructor", this).call(this, elementRef, config);
        this.headerQuery = headerQuery;
        this.radios = [];
        this.id = ++radioGroupIds;
        this.radioIds = -1;
        this.onChange = function (_) {};
        this.onTouched = function (_) {};
        if (ngControl) ngControl.valueAccessor = this;
    }

    _createClass(RadioGroup, [{
        key: "onInit",
        value: function onInit() {
            var header = this.headerQuery.first;
            if (header) {
                if (!header.id) {
                    header.id = 'radio-header-' + this.id;
                }
                this.describedById = header.id;
            }
        }

        /**
         * Register the specified radio button with the radio group.
         * @param {RadioButton} radio  The radio button to register.
         */
    }, {
        key: "registerRadio",
        value: function registerRadio(radio) {
            radio.id = radio.id || 'radio-' + this.id + '-' + ++this.radioIds;
            this.radios.push(radio);
            if (radio.checked) {
                this.value = radio.value;
                this.activeId = radio.id;
            }
        }

        /**
         * Update which radio button in the group is checked, unchecking all others.
         * @param {RadioButton} checkedRadio  The radio button to check.
         */
    }, {
        key: "update",
        value: function update(checkedRadio) {
            this.value = checkedRadio.value;
            this.activeId = checkedRadio.id;
            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
                for (var _iterator = this.radios[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                    var radio = _step.value;

                    radio.checked = radio === checkedRadio;
                }
            } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion && _iterator["return"]) {
                        _iterator["return"]();
                    }
                } finally {
                    if (_didIteratorError) {
                        throw _iteratorError;
                    }
                }
            }

            this.onChange(this.value);
        }

        /**
         * @private
         * Angular2 Forms API method called by the model (Control) on change to update
         * the checked value.
         * https://github.com/angular/angular/blob/master/modules/angular2/src/forms/directives/shared.ts#L34
         */
    }, {
        key: "writeValue",
        value: function writeValue(value) {
            this.value = value;
            var _iteratorNormalCompletion2 = true;
            var _didIteratorError2 = false;
            var _iteratorError2 = undefined;

            try {
                for (var _iterator2 = this.radios[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
                    var radio = _step2.value;

                    radio.checked = radio.value == value;
                }
            } catch (err) {
                _didIteratorError2 = true;
                _iteratorError2 = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
                        _iterator2["return"]();
                    }
                } finally {
                    if (_didIteratorError2) {
                        throw _iteratorError2;
                    }
                }
            }
        }

        /**
         * @private
         * Angular2 Forms API method called by the view (NgControl) to register the
         * onChange event handler that updates the model (Control).
         * https://github.com/angular/angular/blob/master/modules/angular2/src/forms/directives/shared.ts#L27
         * @param {Function} fn  the onChange event handler.
         */
    }, {
        key: "registerOnChange",
        value: function registerOnChange(fn) {
            this.onChange = fn;
        }

        /**
         * @private
         * Angular2 Forms API method called by the the view (NgControl) to register
         * the onTouched event handler that marks the model (Control) as touched.
         * @param {Function} fn  onTouched event handler.
         */
    }, {
        key: "registerOnTouched",
        value: function registerOnTouched(fn) {
            this.onTouched = fn;
        }
    }]);

    return RadioGroup;
})(_ion.Ion);
exports.RadioGroup = RadioGroup;
exports.RadioGroup = RadioGroup = __decorate([(0, _configDecorators.IonicDirective)({
    selector: 'ion-radio-group',
    host: {
        'class': 'list',
        'role': 'radiogroup',
        '[attr.aria-activedescendant]': 'activeId',
        '[attr.aria-describedby]': 'describedById'
    }
}), __param(2, (0, _angular2Angular2.Optional)()), __param(3, (0, _angular2Angular2.Query)(_listList.ListHeader)), __metadata('design:paramtypes', [typeof _angular2Angular2.ElementRef !== 'undefined' && _angular2Angular2.ElementRef || Object, typeof _configConfig.IonicConfig !== 'undefined' && _configConfig.IonicConfig || Object, typeof _angular2Angular2.NgControl !== 'undefined' && _angular2Angular2.NgControl || Object, typeof _angular2Angular2.QueryList !== 'undefined' && _angular2Angular2.QueryList || Object])], RadioGroup);
/**
 * @name ionRadio
 * @description
 * A single radio component.
 *
 * See the [Angular 2 Docs](https://angular.io/docs/js/latest/api/forms/) for more info on forms and input.
 *
 * @usage
 * ```html
 * <ion-radio value="isChecked" checked="true">
 *   Radio Label
 * </ion-radio>
 * ```
 *
 */
var RadioButton = (function (_Ion2) {
    _inherits(RadioButton, _Ion2);

    /**
     * Radio button constructor.
     * @param {RadioGroup=} group  The parent radio group, if any.
     * @param {ElementRef} elementRef  TODO
     * @param {IonicConfig} config  TODO
     */

    function RadioButton(group, elementRef, config) {
        _classCallCheck(this, RadioButton);

        _get(Object.getPrototypeOf(RadioButton.prototype), "constructor", this).call(this, elementRef, config);
        this.group = group;
        this.tabIndex = 0;
    }

    _createClass(RadioButton, [{
        key: "onInit",
        value: function onInit() {
            _get(Object.getPrototypeOf(RadioButton.prototype), "onInit", this).call(this);
            this.group.registerRadio(this);
            this.labelId = 'label-' + this.id;
        }
    }, {
        key: "click",
        value: function click(ev) {
            ev.preventDefault();
            ev.stopPropagation();
            this.check();
        }

        /**
         * Update the checked state of this radio button.
         * TODO: Call this toggle? Since unchecks as well
         */
    }, {
        key: "check",
        value: function check() {
            this.checked = !this.checked;
            this.group.update(this);
        }
    }]);

    return RadioButton;
})(_ion.Ion);
exports.RadioButton = RadioButton;
exports.RadioButton = RadioButton = __decorate([(0, _configDecorators.IonicComponent)({
    selector: 'ion-radio',
    properties: ['value', 'checked', 'disabled', 'id'],
    host: {
        'class': 'item',
        'role': 'radio',
        'tappable': 'true',
        '[attr.id]': 'id',
        '[attr.tab-index]': 'tabIndex',
        '[attr.aria-checked]': 'checked',
        '[attr.aria-disabled]': 'disabled',
        '[attr.aria-labelledby]': 'labelId',
        '(click)': 'click($event)'
    }
}), (0, _configDecorators.IonicView)({
    template: '<ion-item-content id="{{labelId}}">' + '<ng-content></ng-content>' + '</ion-item-content>' + '<media-radio>' + '<radio-icon></radio-icon>' + '</media-radio>'
}), __param(0, (0, _angular2Angular2.Host)()), __param(0, (0, _angular2Angular2.Optional)()), __metadata('design:paramtypes', [RadioGroup, typeof _angular2Angular2.ElementRef !== 'undefined' && _angular2Angular2.ElementRef || Object, typeof _configConfig.IonicConfig !== 'undefined' && _configConfig.IonicConfig || Object])], RadioButton);
var radioGroupIds = -1;
