// ionic-framework/src/es6/ionic/platform/geolocation/geolocation.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import * as Rx from 'rx';
import { NativePlugin } from '../plugin';
/**
 * Get geolocation data.
 *
 * @usage
 * ```js
 * Geolocation.getCurrentPosition().then((resp) => {
 *  //resp.coords.latitude
 *  //resp.coords.longitude
 * })
 *
 * let watch = Geolocation.watchPosition();
 * watch.source.subscribe((data) => {
 *  //data.coords.latitude
 *  //data.coords.longitude
 * })
 * ```
 */
export let Geolocation = class {
    /**
     * Get the current GPS location.
     */
    static getCurrentPosition(options) {
        return new Promise((resolve, reject) => {
            navigator.geolocation.getCurrentPosition(function (result) {
                resolve(result);
            }, function (err) {
                reject(err);
            }, options);
        });
    }
    /**
     * Watch for location changes.
     */
    static watchPosition(options) {
        let watchID;
        let source = Rx.Observable.create((observer) => {
            watchID = navigator.geolocation.watchPosition(function (result) {
                observer.onNext(result);
            }, function (err) {
                observer.onError(err, observer);
            }, options);
        });
        return {
            source: source,
            watchID: watchID,
            clear: () => {
                navigator.geolocation.clearWatch(watchID);
            }
        };
    }
    /**
     * Clear a specific watch by watch ID. Generally, you'll call
     * clear() on the returned watch from `getCurrentPosition` or `watchPosition` above.
     */
    static clearWatch(watchID) {
        return navigator.geolocation.clearWatch(watchID);
    }
};
Geolocation = __decorate([
    NativePlugin({
        name: 'Geolocation',
        platforms: ['ios', 'android', 'web'],
        engines: {
            cordova: 'cordova-plugin-geolocation'
        }
    }), 
    __metadata('design:paramtypes', [])
], Geolocation);
