// ionic-framework/src/es5/common/ionic/platform/applinks/applinks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _plugin = require('../plugin');

/**
 * Open installed apps on the device. Note: Android and iOS have different ways of
 * opening and specifying launch params, so they have separate launch functions.
 *
 * @usage
 *
 * ```js
 * if(platform.is('ios') {
 *   AppLinks.check('twitter://').then((installed) => {
 *     AppLinks.openIOS('twitter://user?screen_name=ionicframework')
 *   }, (err) => {
 *
 *   })
 * } else if(platform.is('android') {
 *   AppLinks.check('com.twitter.android').then((installed) => {
 *     AppLinks.openAndroid([["action", "VIEW"], ['twitter://user?screen_name=ionicframework']])
 *   })
 * }
 * ```
 */
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2:
            return decorators.reduceRight(function (o, d) {
                return d && d(o) || o;
            }, target);
        case 3:
            return decorators.reduceRight(function (o, d) {
                return (d && d(target, key), void 0);
            }, void 0);
        case 4:
            return decorators.reduceRight(function (o, d) {
                return d && d(target, key, o) || o;
            }, desc);
    }
};
var __metadata = undefined && undefined.__metadata || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var AppLinks = (function () {
    function AppLinks() {
        _classCallCheck(this, AppLinks);
    }

    _createClass(AppLinks, null, [{
        key: "openIOS",

        /**
         * Open app on iOS with a given URL (iOS), or scheme (Android)
         */
        value: function openIOS(url) {
            this.ifPlugin(function () {
                navigator.startApp.start(url);
            });
        }

        /**
         * Open app on Android with a given scheme and params.
         */
    }, {
        key: "openAndroid",
        value: function openAndroid(args) {
            this.ifPlugin(function () {
                var _navigator$startApp;

                (_navigator$startApp = navigator.startApp).start.apply(_navigator$startApp, _toConsumableArray(args));
            });
        }

        /**
         * Check if an installed app can be opened from the given URL.
         */
    }, {
        key: "canOpen",
        value: function canOpen(urlOrScheme) {
            var _this = this;

            return new Promise(function (resolve, reject) {
                var hasPlugin = _this.ifPlugin(function () {
                    navigator.startApp.check(urlOrScheme, function (message) {
                        resolve(message);
                    }, function (err) {
                        reject(err);
                    });
                });
                if (!hasPlugin) {
                    reject('Plugin not installed');
                }
            });
        }
    }]);

    return AppLinks;
})();
exports.AppLinks = AppLinks;
exports.AppLinks = AppLinks = __decorate([(0, _plugin.NativePlugin)({
    name: 'AppLinks',
    platforms: ['ios', 'android'],
    engines: {
        cordova: 'cordova-plugin-statusbar'
    },
    pluginCheck: function pluginCheck() {
        return !!navigator.startApp;
    }
}), __metadata('design:paramtypes', [])], AppLinks);
