// ionic-framework/src/es6/ionic/components/nav-bar/nav-bar.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { Component, Directive, Optional, ElementRef, TemplateRef, forwardRef, Inject } from 'angular2/angular2';
import { Ion } from '../ion';
import { ToolbarBase } from '../toolbar/toolbar';
import { IonicConfig } from '../../config/config';
import { IonicView } from '../../config/decorators';
import { IonicApp } from '../app/app';
import { ViewController } from '../nav/view-controller';
import { NavController } from '../nav/nav-controller';
let BackButton = class extends Ion {
    constructor(navCtrl, elementRef, navbar) {
        super(elementRef, null);
        this.navCtrl = navCtrl;
        navbar && navbar.setBackButtonRef(elementRef);
    }
    goBack(ev) {
        ev.stopPropagation();
        ev.preventDefault();
        this.navCtrl && this.navCtrl.pop();
    }
};
BackButton = __decorate([
    Directive({
        selector: '.back-button',
        host: {
            '(click)': 'goBack($event)'
        }
    }),
    __param(0, Optional()),
    __param(2, Optional()),
    __param(2, Inject(forwardRef(() => Navbar))), 
    __metadata('design:paramtypes', [(typeof NavController !== 'undefined' && NavController) || Object, (typeof ElementRef !== 'undefined' && ElementRef) || Object, Navbar])
], BackButton);
let BackButtonText = class extends Ion {
    constructor(elementRef, navbar) {
        super(elementRef, null);
        navbar && navbar.setBackButtonTextRef(elementRef);
    }
};
BackButtonText = __decorate([
    Directive({
        selector: '.back-button-text'
    }),
    __param(1, Optional()),
    __param(1, Inject(forwardRef(() => Navbar))), 
    __metadata('design:paramtypes', [(typeof ElementRef !== 'undefined' && ElementRef) || Object, Navbar])
], BackButtonText);
export let Navbar = class extends ToolbarBase {
    constructor(app, viewCtrl, elementRef, config) {
        super(elementRef, config);
        this.app = app;
        viewCtrl && viewCtrl.navbarView(this);
        this.bbIcon = config.setting('backButtonIcon');
        this.bbDefault = config.setting('backButtonText');
    }
    getBackButtonRef() {
        return this.bbRef;
    }
    setBackButtonRef(backButtonElementRef) {
        this.bbRef = backButtonElementRef;
    }
    getBackButtonTextRef() {
        return this.bbtRef;
    }
    setBackButtonTextRef(backButtonTextElementRef) {
        this.bbtRef = backButtonTextElementRef;
    }
    didEnter() {
        this.app.title(this.getTitleText());
    }
};
Navbar = __decorate([
    Component({
        selector: 'ion-navbar',
        host: {
            'class': 'toolbar'
        }
    }),
    IonicView({
        template: '<div class="toolbar-inner">' +
            '<button class="back-button">' +
            '<icon class="back-button-icon" [name]="bbIcon"></icon>' +
            '<span class="back-button-text">' +
            '<span class="back-default">{{bbDefault}}</span>' +
            '</span>' +
            '</button>' +
            '<ng-content select="[menu-toggle]"></ng-content>' +
            '<ng-content select="ion-title"></ng-content>' +
            '<ng-content select="ion-nav-items[primary]"></ng-content>' +
            '<ng-content select="ion-nav-items[secondary]"></ng-content>' +
            '</div>' +
            '<div class="toolbar-background"></div>',
        directives: [BackButton, BackButtonText]
    }),
    __param(1, Optional()), 
    __metadata('design:paramtypes', [(typeof IonicApp !== 'undefined' && IonicApp) || Object, (typeof ViewController !== 'undefined' && ViewController) || Object, (typeof ElementRef !== 'undefined' && ElementRef) || Object, (typeof IonicConfig !== 'undefined' && IonicConfig) || Object])
], Navbar);
/*
  Used to find and register headers in a view, and this directive's
  content will be moved up to the common navbar location, and created
  using the same context as the view's content area.
*/
export let NavbarTemplate = class {
    constructor(viewCtrl, templateRef) {
        viewCtrl && viewCtrl.addTemplateRef('navbar', templateRef);
    }
};
NavbarTemplate = __decorate([
    Directive({
        selector: 'template[navbar]'
    }),
    __param(0, Optional()),
    __param(1, Optional()), 
    __metadata('design:paramtypes', [(typeof ViewController !== 'undefined' && ViewController) || Object, (typeof TemplateRef !== 'undefined' && TemplateRef) || Object])
], NavbarTemplate);
