// ionic-framework/src/es5/common/ionic/components/slides/slides.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj["default"] = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; }

var _angular2Angular2 = require('angular2/angular2');

var _ion = require('../ion');

var _ionicAnimationsAnimation = require('ionic/animations/animation');

var _ionicGesturesGesture = require('ionic/gestures/gesture');

var _configDecorators = require('../../config/decorators');

var _configConfig = require('../../config/config');

var _ionicUtil = require('ionic/util');

var _utilDom = require('../../util/dom');

var util = _interopRequireWildcard(_ionicUtil);

var _swiperWidget = require('./swiper-widget');

/**
 * Slides is a slide box implementation based on Swiper.js
 *
 * Swiper.js:
 * The most modern mobile touch slider and framework with hardware accelerated transitions
 *
 * http://www.idangero.us/swiper/
 *
 * Copyright 2015, Vladimir Kharlampidi
 * The iDangero.us
 * http://www.idangero.us/
 *
 * Licensed under MIT
 *
 */
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2:
            return decorators.reduceRight(function (o, d) {
                return d && d(o) || o;
            }, target);
        case 3:
            return decorators.reduceRight(function (o, d) {
                return (d && d(target, key), void 0);
            }, void 0);
        case 4:
            return decorators.reduceRight(function (o, d) {
                return d && d(target, key, o) || o;
            }, desc);
    }
};
var __metadata = undefined && undefined.__metadata || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = undefined && undefined.__param || function (paramIndex, decorator) {
    return function (target, key) {
        decorator(target, key, paramIndex);
    };
};
var Slides = (function (_Ion) {
    _inherits(Slides, _Ion);

    /**
     * TODO
     * @param {ElementRef} elementRef  TODO
     */

    function Slides(elementRef, config) {
        var _this = this;

        _classCallCheck(this, Slides);

        _get(Object.getPrototypeOf(Slides.prototype), "constructor", this).call(this, elementRef, config);
        this.rapidUpdate = util.debounce(function () {
            _this.update();
        }, 10);
    }

    _createClass(Slides, [{
        key: "onInit",
        value: function onInit() {
            var _this2 = this;

            if (!this.options) {
                this.options = {};
            }
            this.showPager = util.isTrueProperty(this.pager);
            var options = util.defaults({
                pagination: '.swiper-pagination',
                paginationClickable: true,
                lazyLoading: true,
                preloadImages: false
            }, this.options);
            options.onTap = function (swiper, e) {
                _this2.onTap(swiper, e);
                return _this2.options.onTap && _this2.options.onTap(swiper, e);
            };
            options.onClick = function (swiper, e) {
                _this2.onClick(swiper, e);
                return _this2.options.onClick && _this2.options.onClick(swiper, e);
            };
            options.onDoubleTap = function (swiper, e) {
                _this2.onDoubleTap(swiper, e);
                return _this2.options.onDoubleTap && _this2.options.onDoubleTap(swiper, e);
            };
            options.onTransitionStart = function (swiper, e) {
                _this2.onTransitionStart(swiper, e);
                return _this2.options.onTransitionStart && _this2.options.onTransitionStart(swiper, e);
            };
            options.onTransitionEnd = function (swiper, e) {
                _this2.onTransitionEnd(swiper, e);
                return _this2.options.onTransitionEnd && _this2.options.onTransitionEnd(swiper, e);
            };
            options.onSlideChangeStart = function (swiper) {
                return _this2.options.onSlideChangeStart && _this2.options.onSlideChangeStart(swiper);
            };
            options.onSlideChangeEnd = function (swiper) {
                return _this2.options.onSlideChangeEnd && _this2.options.onSlideChangeEnd(swiper);
            };
            options.onLazyImageLoad = function (swiper, slide, img) {
                return _this2.options.onLazyImageLoad && _this2.options.onLazyImageLoad(swiper, slide, img);
            };
            options.onLazyImageReady = function (swiper, slide, img) {
                return _this2.options.onLazyImageReady && _this2.options.onLazyImageReady(swiper, slide, img);
            };
            var swiper = new _swiperWidget.Swiper(this.getNativeElement().children[0], options);
            this.slider = swiper;
            /*
            * TODO: Finish this
            if(util.isTrueProperty(this.zoom)) {
              this.enableZoom = true;
              setTimeout(() => {
                this.initZoom();
              })
            }
            */
        }
    }, {
        key: "onTap",
        value: function onTap(swiper, e) {}
    }, {
        key: "onClick",
        value: function onClick(swiper, e) {}
    }, {
        key: "onDoubleTap",
        value: function onDoubleTap(swiper, e) {
            this.toggleZoom(swiper, e);
        }
    }, {
        key: "onLazyImageLoad",
        value: function onLazyImageLoad(swiper, slide, img) {}
    }, {
        key: "onLazyImageReady",
        value: function onLazyImageReady(swiper, slide, img) {}

        /*
        nextButton(swiper, e) {
        }
        prevButton() {
        }
        indexButton() {
        }
        */
    }, {
        key: "initZoom",
        value: function initZoom() {
            var _this3 = this;

            this.zoomDuration = this.zoomDuration || 230;
            this.maxScale = this.zoomMax || 3;
            this.zoomElement = this.getNativeElement().children[0].children[0];
            this.zoomElement && this.zoomElement.classList.add('ion-scroll-zoom');
            this.zoomGesture = new _ionicGesturesGesture.Gesture(this.zoomElement);
            this.zoomGesture.listen();
            this.scale = 1;
            this.zoomLastPosX = 0;
            this.zoomLastPosY = 0;
            var last_scale = undefined,
                startX = undefined,
                startY = undefined,
                posX = 0,
                posY = 0,
                zoomRect = undefined;
            this.viewportWidth = this.getNativeElement().offsetWidth;
            this.viewportHeight = this.getNativeElement().offsetHeight;
            this.zoomElement.addEventListener('touchstart', function (e) {
                _this3.onTouchStart(e);
            });
            this.zoomElement.addEventListener('touchmove', function (e) {
                _this3.onTouchMove(e);
            });
            this.zoomElement.addEventListener('touchend', function (e) {
                _this3.onTouchEnd(e);
            });
            this.zoomGesture.on('pinchstart', function (e) {
                last_scale = _this3.scale;
                console.log('Last scale', e.scale);
            });
            this.zoomGesture.on('pinch', function (e) {
                _this3.scale = Math.max(1, Math.min(last_scale * e.scale, 10));
                console.log('Scaling', _this3.scale);
                _this3.zoomElement.style[_utilDom.CSS.transform] = 'scale(' + _this3.scale + ')';
                zoomRect = _this3.zoomElement.getBoundingClientRect();
            });
            this.zoomGesture.on('pinchend', function (e) {
                //last_scale = Math.max(1, Math.min(last_scale * e.scale, 10));
                if (_this3.scale > _this3.maxScale) {
                    var za = new _ionicAnimationsAnimation.Animation(_this3.zoomElement).duration(_this3.zoomDuration).easing('linear').from('scale', _this3.scale).to('scale', _this3.maxScale);
                    za.play();
                    _this3.scale = _this3.maxScale;
                }
            });
        }
    }, {
        key: "resetZoom",
        value: function resetZoom() {
            if (this.zoomElement) {
                this.zoomElement.parentElement.style[_utilDom.CSS.transform] = '';
                this.zoomElement.style[_utilDom.CSS.transform] = 'scale(1)';
            }
            this.scale = 1;
            this.zoomLastPosX = 0;
            this.zoomLastPosY = 0;
        }
    }, {
        key: "toggleZoom",
        value: function toggleZoom(swiper, e) {
            console.log('Try toggle zoom');
            if (!this.enableZoom) {
                return;
            }
            console.log('Toggling zoom', e);
            /*
            let x = e.pointers[0].clientX;
            let y = e.pointers[0].clientY;
                 let mx = this.viewportWidth / 2;
            let my = this.viewportHeight / 2;
                 let tx, ty;
                 if(x > mx) {
              // Greater than half
              tx = -x;
            } else {
              // Less than or equal to half
              tx = (this.viewportWidth - x);
            }
            if(y > my) {
              ty = -y;
            } else {
              ty = y-my;
            }
                 console.log(y);
            */
            var zi = new _ionicAnimationsAnimation.Animation(this.touch.target.children[0]).duration(this.zoomDuration).easing('linear').fill('none');
            var zw = new _ionicAnimationsAnimation.Animation(this.touch.target.children[0]).duration(this.zoomDuration).easing('linear');
            var za = new _ionicAnimationsAnimation.Animation();
            za.fill('none');
            za.add(zi); //, zw);
            if (this.scale > 1) {
                // Zoom out
                //zw.fromTo('translateX', posX + 'px', '0px');
                //zw.fromTo('translateY', posY + 'px', '0px');
                zi.from('scale', this.scale);
                zi.to('scale', 1);
                za.play();
                //posX = 0;
                //posY = 0;
                this.scale = 1;
            } else {
                // Zoom in
                //zw.fromTo('translateX', posX + 'px', tx + 'px');
                //zw.fromTo('translateY', posY + 'px', ty + 'px');
                zi.from('scale', this.scale);
                zi.to('scale', this.maxScale);
                za.play();
                //posX = tx;
                //posY = ty;
                this.scale = this.maxScale;
            }
        }
    }, {
        key: "onTransitionStart",
        value: function onTransitionStart(swiper) {}
    }, {
        key: "onTransitionEnd",
        value: function onTransitionEnd(swiper) {}
    }, {
        key: "onTouchStart",
        value: function onTouchStart(e) {
            console.log('Touch start', e);
            //TODO: Support mice as well
            var target = _ionicUtil.dom.closest(e.target, '.slide').children[0].children[0];
            this.touch = {
                startX: e.touches[0].clientX,
                startY: e.touches[0].clientY,
                deltaX: 0,
                deltaY: 0,
                lastX: 0,
                lastY: 0,
                target: target.parentElement,
                zoomable: target,
                zoomableWidth: target.offsetWidth,
                zoomableHeight: target.offsetHeight
            };
            console.log('Target', this.touch.target);
            //TODO: android prevent default
        }
    }, {
        key: "onTouchMove",
        value: function onTouchMove(e) {
            this.touch.deltaX = e.touches[0].clientX - this.touch.startX;
            this.touch.deltaY = e.touches[0].clientY - this.touch.startY;
            // TODO: Make sure we need to transform (image is bigger than viewport)
            var zoomableScaledWidth = this.touch.zoomableWidth * this.scale;
            var zoomableScaledHeight = this.touch.zoomableHeight * this.scale;
            var x1 = Math.min(this.viewportWidth / 2 - zoomableScaledWidth / 2, 0);
            var x2 = -x1;
            var y1 = Math.min(this.viewportHeight / 2 - zoomableScaledHeight / 2, 0);
            var y2 = -y1;
            console.log('BOUNDS', x1, x2, y1, y2);
            if (this.scale <= 1) {
                return;
            }
            console.log('PAN', e);
            // Move image
            this.touch.x = this.touch.deltaX + this.touch.lastX;
            this.touch.y = this.touch.deltaY + this.touch.lastY;
            console.log(this.touch.x, this.touch.y);
            if (this.touch.x < x1) {
                console.log('OUT ON LEFT');
            }
            if (this.touch.x > x2) {
                console.log('OUT ON RIGHT');
            }
            if (this.touch.x > this.viewportWidth) {} else if (-this.touch.x > this.viewportWidth) {} else {
                console.log('TRANSFORM', this.touch.x, this.touch.y, this.touch.target);
                //this.touch.target.style[CSS.transform] = 'translateX(' + this.touch.x + 'px) translateY(' + this.touch.y + 'px)';
                this.touch.target.style[_utilDom.CSS.transform] = 'translateX(' + this.touch.x + 'px) translateY(' + this.touch.y + 'px)';
                e.preventDefault();
                e.stopPropagation();
                return false;
            }
        }
    }, {
        key: "onTouchEnd",
        value: function onTouchEnd(e) {
            console.log('PANEND', e);
            if (this.scale > 1) {
                if (Math.abs(this.touch.x) > this.viewportWidth) {
                    posX = posX > 0 ? this.viewportWidth - 1 : -(this.viewportWidth - 1);
                    console.log('Setting on posx', this.touch.x);
                }
                /*
                if(posY > this.viewportHeight/2) {
                  let z = new Animation(this.zoomElement.parentElement);
                  z.fromTo('translateY', posY + 'px', Math.min(this.viewportHeight/2 + 30, posY));
                  z.play();
                } else {
                  let z = new Animation(this.zoomElement.parentElement);
                  z.fromTo('translateY', posY + 'px', Math.max(this.viewportHeight/2 - 30, posY));
                  z.play();
                }
                */
                this.touch.lastX = this.touch.x;
                this.touch.lastY = this.touch.y;
            }
        }

        /**
         * Update the underlying slider implementation. Call this if you've added or removed
         * child slides.
         */
    }, {
        key: "update",
        value: function update() {
            var _this4 = this;

            setTimeout(function () {
                _this4.slider.update();
                // Don't allow pager to show with > 10 slides
                if (_this4.slider.slides.length > 10) {
                    _this4.showPager = false;
                }
            });
        }
    }, {
        key: "next",
        value: function next() {
            this.slider.slideNext();
        }
    }, {
        key: "prev",
        value: function prev() {
            this.slider.slidePrev();
        }
    }, {
        key: "getIndex",
        value: function getIndex() {
            return this.slider.activeIndex;
        }
    }, {
        key: "getNumSlides",
        value: function getNumSlides() {
            return this.slider.slides.length;
        }
    }, {
        key: "isAtEnd",
        value: function isAtEnd() {
            return this.slider.isEnd;
        }
    }, {
        key: "isAtBeginning",
        value: function isAtBeginning() {
            return this.slider.isBeginning;
        }
    }, {
        key: "getSliderWidget",
        value: function getSliderWidget() {
            return this.slider;
        }
    }]);

    return Slides;
})(_ion.Ion);
exports.Slides = Slides;
exports.Slides = Slides = __decorate([(0, _configDecorators.IonicComponent)({
    selector: 'ion-slides',
    properties: ['loop', 'index', 'bounce', 'pager', 'options', 'zoom', 'zoomDuration', 'zoomMax']
}), (0, _angular2Angular2.View)({
    template: "<div class=\"swiper-container\">\n    <div class=\"swiper-wrapper\">\n      <ng-content></ng-content>\n    </div>\n    <div [class.hide]=\"!showPager\" class=\"swiper-pagination\"></div>\n  </div>",
    directives: [_angular2Angular2.NgIf, _angular2Angular2.NgClass]
}), __metadata('design:paramtypes', [typeof _angular2Angular2.ElementRef !== 'undefined' && _angular2Angular2.ElementRef || Object, typeof _configConfig.IonicConfig !== 'undefined' && _configConfig.IonicConfig || Object])], Slides);
/**
 * TODO
 */
var Slide =
/**
 * TODO
 * @param {Slides} slides  The containing slidebox.
 * @param {ElementRef} elementRef  TODO
 */
function Slide(elementRef, slides) {
    _classCallCheck(this, Slide);

    this.ele = elementRef.nativeElement;
    this.ele.classList.add('swiper-slide');
    slides.rapidUpdate();
};
exports.Slide = Slide;
exports.Slide = Slide = __decorate([(0, _configDecorators.IonicComponent)({
    selector: 'ion-slide',
    properties: ['zoom']
}), (0, _angular2Angular2.View)({
    template: "<div class=\"slide-zoom\"><ng-content></ng-content></div>"
}), __param(1, (0, _angular2Angular2.Host)()), __metadata('design:paramtypes', [typeof _angular2Angular2.ElementRef !== 'undefined' && _angular2Angular2.ElementRef || Object, Slides])], Slide);
var SlideLazy = function SlideLazy(elementRef) {
    _classCallCheck(this, SlideLazy);

    elementRef.getNativeElement().classList.add('swiper-lazy');
};
exports.SlideLazy = SlideLazy;
exports.SlideLazy = SlideLazy = __decorate([(0, _configDecorators.IonicDirective)({
    selector: 'slide-lazy'
}), __metadata('design:paramtypes', [typeof _angular2Angular2.ElementRef !== 'undefined' && _angular2Angular2.ElementRef || Object])], SlideLazy);
