// ionic-framework/src/es5/common/ionic/components/app/id.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _angular2Angular2 = require('angular2/angular2');

var _app = require('./app');

/**
 * IdRef is an easy way to identify unique components in an app and access them
 * no matter where in the UI heirarchy you are. For example, this makes toggling
 * a global side menu feasible from any place in the application.
 */
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2:
            return decorators.reduceRight(function (o, d) {
                return d && d(o) || o;
            }, target);
        case 3:
            return decorators.reduceRight(function (o, d) {
                return (d && d(target, key), void 0);
            }, void 0);
        case 4:
            return decorators.reduceRight(function (o, d) {
                return d && d(target, key, o) || o;
            }, desc);
    }
};
var __metadata = undefined && undefined.__metadata || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var IdRef = (function () {
    function IdRef(app, elementRef, appViewManager) {
        _classCallCheck(this, IdRef);

        this.app = app;
        this.elementRef = elementRef;
        this.appViewManager = appViewManager;
        // Grab the component this directive is attached to
        this.component = appViewManager.getComponent(elementRef);
    }

    _createClass(IdRef, [{
        key: "onInit",
        value: function onInit() {
            this.app.register(this.id, this.component);
        }
    }, {
        key: "onDestroy",
        value: function onDestroy() {
            this.app.unregister(this.id);
        }
    }]);

    return IdRef;
})();
exports.IdRef = IdRef;
exports.IdRef = IdRef = __decorate([(0, _angular2Angular2.Directive)({
    selector: '[id]',
    properties: ['id']
}), __metadata('design:paramtypes', [typeof _app.IonicApp !== 'undefined' && _app.IonicApp || Object, typeof _angular2Angular2.ElementRef !== 'undefined' && _angular2Angular2.ElementRef || Object, typeof _angular2Angular2.AppViewManager !== 'undefined' && _angular2Angular2.AppViewManager || Object])], IdRef);
