// ionic-framework/src/es6/ionic/components/show-hide-when/show-hide-when.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { Directive, Attribute, NgZone } from 'angular2/angular2';
import { IonicPlatform } from '../../platform/platform';
class DisplayWhen {
    constructor(conditions, platform, ngZone) {
        this.isMatch = false;
        this.platform = platform;
        if (!conditions)
            return;
        this.conditions = conditions.split(',');
        // check if its one of the matching platforms first
        // a platform does not change during the life of an app
        for (let i = 0; i < this.conditions.length; i++) {
            if (this.conditions[i] && platform.is(this.conditions[i])) {
                this.isMatch = true;
                return;
            }
        }
        if (this.orientation()) {
            // add window resize listener
            platform.onResize(() => {
                ngZone.run(() => {
                    this.orientation();
                });
            });
            return;
        }
    }
    orientation() {
        for (let i = 0; i < this.conditions.length; i++) {
            if (this.conditions[i] == 'portrait') {
                this.isMatch = this.platform.isPortrait();
                return true;
            }
            if (this.conditions[i] == 'landscape') {
                this.isMatch = this.platform.isLandscape();
                return true;
            }
        }
    }
}
/**
 * TODO
 */
export let ShowWhen = class extends DisplayWhen {
    /**
     * TODO
     * @param {string} showWhen  The value of the element's 'show-when' attribute
     * @param {NgZone} ngZone  TODO
     */
    constructor(showWhen, platform, ngZone) {
        super(showWhen, platform, ngZone);
    }
    get hidden() {
        return !this.isMatch;
    }
};
ShowWhen = __decorate([
    Directive({
        selector: '[show-when]',
        host: {
            '[hidden]': 'hidden'
        }
    }),
    __param(0, Attribute('show-when')), 
    __metadata('design:paramtypes', [String, (typeof IonicPlatform !== 'undefined' && IonicPlatform) || Object, (typeof NgZone !== 'undefined' && NgZone) || Object])
], ShowWhen);
/**
 * TODO
 */
export let HideWhen = class extends DisplayWhen {
    /**
     * TODO
     * @param {string} showWhen  The value of the element's 'hide-when' attribute
     * @param {NgZone} ngZone  TODO
     */
    constructor(hideWhen, platform, ngZone) {
        super(hideWhen, platform, ngZone);
    }
    get hidden() {
        return this.isMatch;
    }
};
HideWhen = __decorate([
    Directive({
        selector: '[hide-when]',
        host: {
            '[hidden]': 'hidden'
        }
    }),
    __param(0, Attribute('hide-when')), 
    __metadata('design:paramtypes', [String, (typeof IonicPlatform !== 'undefined' && IonicPlatform) || Object, (typeof NgZone !== 'undefined' && NgZone) || Object])
], HideWhen);
